components/webext-storage/src/sync/outgoing.rs

```diff
@@ -0,0 +1,147 @@
+// managing the sync state of the local DB.
+
+use interrupt::Interruptee;
+use rusqlite::{Connection, Row};
+use sql_support::ConnExt;
+use sync15::ServerTimestamp;
+use sync_guid::Guid as SyncGuid;
+
+use crate::error::*;
+
+use super::{ServerPayload, SyncStatus};
+
+// This is the "state" for outgoing items - it's so that after we POST the
+// outgoing records we can update the local DB.
+pub struct OutgoingStateHolder {
+    ext_id: String,
+    change_counter: i32,
+}
+
+pub struct OutgoingInfo {
+    state: OutgoingStateHolder,
+    payload: ServerPayload,
+}
+
+impl OutgoingInfo {
+    fn from_row(row: &Row<'_>) -> Result<Self> {
+        let guid = row
+            .get::<_, Option<SyncGuid>>("guid")?
+            .unwrap_or_else(SyncGuid::random);
+        let ext_id: String = row.get("ext_id")?;
+        let raw_data: Option<String> = row.get("data")?;
+        let (data, deleted) = if raw_data.is_some() {
+            (raw_data, false)
+        } else {
+            (None, true)
+        };
+        Ok(OutgoingInfo {
+            state: OutgoingStateHolder {
+                ext_id: ext_id.clone(),
+                change_counter: row.get("sync_change_counter")?,
+            },
+            payload: ServerPayload {
+                ext_id: ext_id,
+                guid,
+                data,
+                deleted,
+                last_modified: ServerTimestamp(0),
+            },
+        })
+    }
+}
+
+/// Gets into about what should be uploaded. Returns a vec of the payload which
+// should be uploaded, plus the state for those items which should be held
+// until the upload is complete, then passed back to record_uploaded.
+pub fn get_outgoing<S: ?Sized + Interruptee>(
+    conn: &Connection,
+    _signal: &S,
+) -> Result<Vec<OutgoingInfo>> {
+    let sql = "SELECT l.ext_id, l.data, l.sync_change_counter, m.guid
+               FROM moz_extension_data l
+               LEFT JOIN moz_extension_data_mirror m ON m.ext_id = l.ext_id
+               WHERE sync_change_counter > 0";
+    let elts = conn
+        .conn()
+        .query_rows_and_then_named(sql, &[], OutgoingInfo::from_row)?;
+    Ok(elts)
+}
+
+pub fn record_uploaded<S: ?Sized + Interruptee>(
+    conn: &Connection,
+    items: &[&OutgoingStateHolder],
+    signal: &S,
+) -> Result<()> {
+    let cext = conn.conn();
+    let tx = cext.unchecked_transaction()?;
+
+    let sql = "
+        UPDATE moz_extension_data SET
+            sync_change_counter = (sync_change_counter - :old_counter),
+            sync_status = :sync_status_normal
+        WHERE ext_id = :ext_id;";
+    for state in items.iter() {
+        signal.err_if_interrupted()?;
+        conn.execute_named(
+            sql,
+            rusqlite::named_params! {
+                ":old_counter": state.change_counter,
+                ":sync_status_normal": SyncStatus::Normal as u8,
+                ":ext_id": state.ext_id,
+            },
+        )?;
+    }
```

Comment:
I understand this code isn't all plumbed together yet, but would I be correct to expect that this function should also update the mirror to reflect what was uploaded to the server? Or, will we sync down our own changes and only update the mirror at that point?

Reply:
yes, exactly. That's in a later version too.